compiler/src/linking/link.re

```diff
@@ -447,35 +456,59 @@ let link_all = (linked_mod, dependencies, signature) => {
             let value =
               switch (imported_name) {
               | "relocBase" =>
-                Expression.Const.make(
-                  wasm_mod,
-                  Literal.int32(Int32.of_int(table_offset^)),
+                Some(
+                  Expression.Const.make(
+                    wasm_mod,
+                    Literal.int32(Int32.of_int(table_offset^)),
+                  ),
                 )
               | "moduleRuntimeId" =>
                 // Module id is tagged; incrementing by 2 is the equivalent of an untagged increment by 1
                 module_id := module_id^ + 2;
-                Expression.Const.make(
-                  wasm_mod,
-                  Literal.int32(Int32.of_int(module_id^)),
+                Some(
+                  Expression.Const.make(
+                    wasm_mod,
+                    Literal.int32(Int32.of_int(module_id^)),
+                  ),
                 );
               | "runtimeHeapStart" =>
-                let size =
-                  Option.value(
-                    ~default=0,
-                    Option.map(Bytes.length, metadata_tbl_data),
-                  );
-                let runtime_heap_start = metadata_heap_loc + size;
-                Expression.Const.make(
-                  wasm_mod,
-                  Literal.int32(Int32.of_int(runtime_heap_start)),
+                Some(
+                  Expression.Const.make(
+                    wasm_mod,
+                    Literal.int32(Int32.of_int(runtime_heap_start)),
+                  ),
+                )
+              | "runtimeHeapNextPtr" =>
+                Hashtbl.replace(
+                  global_names,
+                  internal_name,
+                  runtime_heap_next_ptr,
                 );
+                None;
+              | "metadataPtr" =>
+                Some(
+                  Expression.Const.make(
+                    wasm_mod,
+                    Literal.int32(Int32.of_int(metadata_heap_loc)),
+                  ),
+                )
               | value =>
                 failwith(
```

Comment:
If you are changing the above to `Some()`, why not make this a None instead of a failwith?

Reply:
`None` would not be correct here... We truly want to throw an error because something is horribly broken in this case. We wouldn't want to silently pass on it.